src/App.js

```diff
@@ -23,49 +23,127 @@ const generateSquares = () => {
+  // You will need to create a method to change the square
   //   When it is clicked on.
   //   Then pass it into the squares as a callback
 
-
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if 
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if 
-    //    all three squares have the same value.
+    let i = 0;
 
-  }
+    // Check all the rows and columns for a winner
+    while (i < 3) {
+      if (
+        squares[i][0].value === squares[i][1].value &&
+        squares[i][2].value === squares[i][1].value &&
+        squares[i][0].value !== ''
+      ) {
+        return squares[i][0].value;
+      } else if (
+        squares[0][i].value === squares[1][i].value &&
+        squares[2][i].value === squares[1][i].value &&
+        squares[0][i].value !== ''
+      ) {
+        return squares[0][i].value;
+      }
+      i += 1;
+    }
+    // Check Top-Left to bottom-right diagonal
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][0].value;
+    }
+
+    // Check Top-right to bottom-left diagonal
+    if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][2].value;
+    }
+
+    return null;
+  };
 
   const resetGame = () => {
-    // Complete in Wave 4
+    let resetBoard = generateSquares();
+    setCurrentPlayer(PLAYER_1);
+    setWinner(null);
+    setSquares(resetBoard);
+  };
+
+  const onClickCallback = (id) => {
+    console.log('onclick', id);
+    setSquares((squares) => {
+      let newBoard = squares.map((square) => {
+        for (let property of square) {
+          if (property.id === id && property.value === '') {
+            if (currentPlayer === PLAYER_1) {
+              property.value = PLAYER_1;
+            } else if (currentPlayer === PLAYER_2) {
+              property.value = PLAYER_2;
+            }
+          }
+        }
+
+        return square;
+      });
+
+      setWinner(checkForWinner());
+      return newBoard;
+    });
+
+    if (currentPlayer === PLAYER_1) {
+      setCurrentPlayer(PLAYER_2);
+    } else {
+      setCurrentPlayer(PLAYER_1);
+    }
+  };
+
+  let header;
+  let boardCallback;
+  // logic to display 'Player 1' instead of x, disabled for tests
+  // let finalWinner;
+  if (winner != null) {
+    // if (winner === PLAYER_1) {
+    //   finalWinner = 'Player 1';
+    // } else {
+    //   finalWinner = 'Player 2';
+    // }
+    header = <h2>Winner is {winner}</h2>;
+  } else {
+    header = <h2>The Current Player is {currentPlayer}</h2>;
+    boardCallback = onClickCallback;
```

Comment:
Love this! I think what could make this just a bit cleaner would be to make it a helper function and make `header` a piece of state. Any changing variables should be in state because things like `useEffect` and other hooks only listen to state, not variables a programmer might make.

So here, doesn't give it your JSX any issue, but in the future it's a good rule of thumb!